### PR TITLE
i#5651: Switch DR's suspend signal to SIGILL with sigqueueinfo

### LINKED ISSUE #5651: gdb has poor support for DR's new takeover signal SIGSTKFLT
For #5458 we switched from SIGUSR2 to SIGSTKFLT.  However, gdb does not handle this signal well at all:

```
Thread 2 "tool.drcacheoff" received signal ?, Unknown signal.
(gdb) p $_siginfo
$1 = {
  si_signo = 16,
(gdb) c
Continuing.
warning: Signal ? does not exist on this system.

(gdb) handle 16 nostop noprint pass
Only signals 1-15 are valid as numeric signals.
Use "info signals" for a list of symbolic signals.
(gdb) handle SIGSTKFLT nostop noprint pass
Unrecognized or ambiguous flag word: "SIGSTKFLT".
```

I tried `handle all nostop noprint pass` and it didn't apply to signal 16!
So we have to hit `c` a hundred times to make progress which is not good.

I saw this in gdb version 9 and 10.

I propose we switch to SIGPWR which gdb does support.  I tested it with gdb and with QEMU and QEMU does not freak out and abort like it does with SIGFPE.

Other candidates are SIGXCPU and SIGXFSZ but the kernel will produce those: SIGPWR seems more obscure.

### PR DESCRIPTION
**Pull request by derekbruening**
Switches the signal DR uses to suspend threads from SIGSTKFLT, which is not supported in gdb, to SIGPWR.  SIGPWR is similarly obscure and never used on Linux, but is supported by gdb, and does not cause QEMU to fail.

Fixes #5651

### CODE REVIEW

**Comment by derekbruening** [comment]
Still hitting problems with QEMU limitations: it seems to not honor our siginfo fields sent through SYS_rt_tgsigqueueinfo, clearing si_errno!  Once again going to have to use a special workaround.

**Comment by derekbruening** [comment]
@abhinav92003 PTAL

File: api/docs/debugging.dox [hunk 1]
```diff
@@ -78,12 +78,10 @@ the shared library that identifies the debug file.
 
 ## Expected Signals
 
-On Linux, DR sends itself a SIGILL signal during initialization, in order to measure the size of the signal frame used by the kernel.  If you are under gdb, simply continue past this signal.
+On Linux, DR sends itself a SIGILL signal during initialization, in order to measure the size of the signal frame used by the kernel.  If you are under gdb, simply continue past this signal.  SIGILL (SIGFPE on MacOS; SIGSTKFLT inside QEMU) is also sent to other threads to suspend or terminate them later on during execution.
```
Comment (abhinav92003) [inline api/docs/debugging.dox:81]: Now SIGILL is used also for attach right?

File: core/unix/nudgesig.c [hunk 2]
```diff
@@ -50,7 +50,7 @@
 
 /* shared with tools/nudgeunix.c */
 bool
-create_nudge_signal_payload(kernel_siginfo_t *info OUT, uint action_mask,
+create_nudge_signal_payload(kernel_siginfo_t *info OUT, uint action_mask, uint flags,
                             client_id_t client_id, uint64 client_arg)
 {
```
Comment (abhinav92003) [inline core/unix/nudgesig.c:55]: Assert that flags is not more than 2 bits.

File: core/lib/globals_shared.h [hunk 3]
```diff
@@ -928,6 +928,7 @@ typedef enum {
  */
 enum {
     NUDGE_IS_INTERNAL = 0x01, /* nudge is internally generated */
+    NUDGE_IS_SUSPEND = 0x02,  /* This is an internal SUSPEND_SIGNAL. */
 #ifdef WINDOWS
     NUDGE_NUDGER_FREE_STACK = 0x02, /* nudger will free the nudge thread's stack so the
                                      * nudge thread itself shouldn't */
```
Comment (abhinav92003) [inline core/lib/globals_shared.h:934]: This goes into nudge_arg_t.flags right? But that field is only 2 bits and here we have 0x04 which won't fit.

File: core/unix/nudgesig.c [hunk 4]
```diff
@@ -80,7 +81,7 @@ send_nudge_signal(process_id_t pid, uint action_mask, client_id_t client_id,
...
     kernel_siginfo_t info;
     int res;
-    if (!create_nudge_signal_payload(&info, action_mask, client_id, client_arg))
+    if (!create_nudge_signal_payload(&info, action_mask, 0, client_id, client_arg))
```
Comment (abhinav92003) [inline core/unix/nudgesig.c:84]: There's no value corresponding to zero in nudge_arg_t flags.

File: api/docs/debugging.dox (hunk 1 above)
Comment (abhinav92003) [inline api/docs/debugging.dox:81]: Also document lack of gdb support on qemu.

File: core/unix/os.c [hunk 5]
```diff
@@ -3714,6 +3758,34 @@ os_thread_sleep(uint64 milliseconds)
...
+    if (is_sigqueueinfo_enosys) {
+        /* We won't be able to distinguish a suspend from a nudge, but we don't
+         * support nudges on old Linux kernels anyway (<2.6.31).
+         */
+        if (tr != NULL)
+            return known_thread_signal(tr, SUSPEND_SIGNAL);
+        else
+            return thread_signal(pid, tid, SUSPEND_SIGNAL);
+    }
+    kernel_siginfo_t info;
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply core/unix/os.c:3780]: I see we always set is_sigqueueinfo_enosys for QEMU, so we won't reach here. Add a comment here.

File: core/unix/signal.c [hunk 6]
```diff
@@ -6037,6 +6020,22 @@ main_signal_handler_C(byte *xsp)
 #endif
 
     default: {
+        /* Handle nudges and suspends, which may use the same signal. */
+        if (sig == SUSPEND_SIGNAL) {
+            if (handle_suspend_signal(dcontext, siginfo, ucxt, frame)) {
+                /* i#1921: see comment above */
+                ASSERT(tr == NULL || tr->under_dynamo_control ||
```
Comment (abhinav92003) [inline core/unix/signal.c:6027]: Is this assert valid for nudges too?

File: core/unix/signal.c [hunk 7]
```diff
@@ -8329,6 +8328,39 @@ sig_detach(dcontext_t *dcontext, sigframe_rt_t *frame, KSYNCH_TYPE *detached)
...
+    if (siginfo->si_signo != sig_expecting)
+        return false;
+    if (is_sigqueue_supported()) {
+        if (siginfo->si_code != SI_QUEUE || siginfo->si_errno == 0)
+            return false;
+    } else {
+        /* Distinguish from a synchronous app signal by looking for
+         * si_code where <=0 means user-sent.
+         */
+        if (siginfo->si_code > 0)
```
Comment (abhinav92003) [inline core/unix/signal.c:8358]: So our signal will have > 0? Do we set it ourselves? if not, how is it set?

File: core/unix/signal.c [hunk 8]
```diff
@@ -8339,15 +8371,20 @@ handle_suspend_signal(dcontext_t *dcontext, kernel_siginfo_t *siginfo,
...
-     * We distinguish further below from the rare case of an app sending
-     * SUSPEND_SIGNAL asynchronously by looking for our particular state settings
-     * that correspond to the use of this signal by DR..
-     */
-    if (siginfo->si_code > 0 || siginfo->si_signo != SUSPEND_SIGNAL)
+    if (!is_signal_for_us(siginfo, SUSPEND_SIGNAL))
         return true; /* pass to app */
 
+    if (is_sigqueue_supported() && SUSPEND_SIGNAL == NUDGESIG_SIGNUM) {
+        nudge_arg_t *arg = (nudge_arg_t *)siginfo;
```
Comment (abhinav92003) [inline core/unix/signal.c:8378]: Should we further check is_signal_for_us(siginfo, NUDGESIG_SIGNUM)?

File: tools/nudgeunix.c [hunk 9]
```diff
@@ -140,7 +140,7 @@ main(int argc, const char *argv[])
         return usage();
 
     /* construct the payload */
-    success = create_nudge_signal_payload(&info, action_mask, client_id, client_arg);
+    success = create_nudge_signal_payload(&info, action_mask, 0, client_id, client_arg);
```
Comment (abhinav92003) [inline tools/nudgeunix.c:143]: Is zero a proper value in the flags enum?

**Review by abhinav92003 — APPROVED** [review]

File: core/unix/os.c [hunk 10]
```diff
@@ -514,8 +514,11 @@ typedef int *(*errno_loc_t)(void);
 #ifdef LINUX
 /* Stores whether clone3 is unsupported on the system we're running on. */
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/os.c:515]: Done.

File: core/unix/os.c [hunk 11]
```diff
@@ -892,9 +895,27 @@ detect_unsupported_syscalls()
         dynamorio_syscall(SYS_clone3, 2, NULL /*clone_args*/, 0 /*clone_args_size*/);
     ASSERT(clone3_errno == -ENOSYS || clone3_errno == -EINVAL);
     is_clone3_enosys = clone3_errno == -ENOSYS;
+    int sigqueue_errno = dynamorio_syscall(SYS_rt_tgsigqueueinfo, 4, get_process_id(),
+                                           get_sys_thread_id(), -1, NULL);
+    ASSERT(sigqueue_errno == -ENOSYS || sigqueue_errno == -EINVAL ||
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/os.c:900]: The NULL raises EFAULT but you could imagine a kernel giving EINVAL for the -1; really we just want it to fail so either way is fine.  Added a comment.

File: core/unix/os.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/os.c:3780]: Even if the suspend sig != nudge sig, we would prefer to use the payload to better distinguish a suspend sig from an app using that sig: the only reason we wouldn't want to get here is if we can't use sigqueueinfo for some reason, which the if() asking whether it's available seems clear enough I would think.

File: core/unix/signal.c [hunk 12]
```diff
@@ -6037,6 +6020,22 @@ main_signal_handler_C(byte *xsp)
...
+        if (sig == SUSPEND_SIGNAL) {
+            if (handle_suspend_signal(dcontext, siginfo, ucxt, frame)) {
+                /* i#1921: see comment above */
+                ASSERT(tr == NULL || tr->under_dynamo_control ||
+                       IS_CLIENT_THREAD(dcontext));
+                record_pending_signal(dcontext, sig, ucxt, frame, false, NULL);
+            }
+            /* else, don't deliver to app */
+            break;
+        } else if (sig == NUDGESIG_SIGNUM) {
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/signal.c:6033]: That feels less safe as a change to how signals are delivered to not use a break would then result in double-handling the same signal.

File: core/unix/signal.c [hunk 13]
```diff
@@ -6037,6 +6020,22 @@ main_signal_handler_C(byte *xsp)
 #endif
 
     default: {
+        /* Handle nudges and suspends, which may use the same signal. */
+        if (sig == SUSPEND_SIGNAL) {
+            if (handle_suspend_signal(dcontext, siginfo, ucxt, frame)) {
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/signal.c:6025]: The comment already there at line 6023 says "Handle nudges and suspends, which may use the same signal."  Added another comment at the else.

> Also currently we're taking the same action (record_pending_signal) for nudges and suspend,

That's not the action for an actual nudge or suspend: that's if this signal is *not* one of those and is instead meant for the app, which is what we do for all signals for the app.

File: core/unix/nudgesig.c [hunk 14]
```diff
@@ -62,12 +62,13 @@ create_nudge_signal_payload(kernel_siginfo_t *info OUT, uint action_mask,
     arg = (nudge_arg_t *)info;
     arg->version = NUDGE_ARG_CURRENT_VERSION;
     arg->nudge_action_mask = action_mask;
-    arg->flags = 0;
+    arg->flags = flags;
     arg->client_id = client_id;
     arg->client_arg = client_arg;
 
     /* ensure nudge_arg_t overlays how we expect it to */
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/nudgesig.c:69]: This is used in several different libraries with different error reporting schemes: I think it is better to let them handle the failure which is the return value here.

File: api/docs/debugging.dox (hunk 1 above)
Comment (derekbruening) [reply api/docs/debugging.dox:81]: See the next sentence "SIGILL...is also sent..."

File: core/unix/nudgesig.c (hunk 2 above)
Comment (derekbruening) [reply core/unix/nudgesig.c:55]: Done.

File: core/lib/globals_shared.h (hunk 3 above)
Comment (derekbruening) [reply core/lib/globals_shared.h:934]: The flags field is 32 bits on Windows.

Comment (derekbruening) [reply core/lib/globals_shared.h:934]: (There is already a reminder comment for the 2 bits just on Linux right above this enum)

File: api/docs/debugging.dox (hunk 1 above)
Comment (derekbruening) [reply api/docs/debugging.dox:81]: Added a mention.  There are many issues with qemu and I think using gdb on DR on QEMU is already not very workable even w/o SIGSTKFLT.

File: core/unix/signal.c (hunk 6 above)
Comment (derekbruening) [reply core/unix/signal.c:6027]: Yes, this is already asserted at line 5739 for all signals but with an extra || sig == SUSPEND_SIGNAL b/c we use that on unknown threads.

File: core/unix/signal.c (hunk 7 above)
Comment (derekbruening) [reply core/unix/signal.c:8358]: The kernel sets this field.

File: core/unix/signal.c (hunk 8 above)
Comment (derekbruening) [reply core/unix/signal.c:8378]: But if they're the same number we've already called that.

File: tools/nudgeunix.c (hunk 9 above)
Comment (derekbruening) [reply tools/nudgeunix.c:143]: See earlier reply: this is not an enum typed parameter but an integer.

**Comment by derekbruening** [comment]
a64 hit windows-zlib #5493
win64 hit #5655

**Pull request merged**